fairlearn/metrics/_metrics_engine.py

```diff
@@ -49,29 +55,57 @@ def metric_by_group(metric_function,
     if sample_weight is not None:
         s_w = _convert_to_ndarray_and_squeeze(sample_weight)
 
-    # Evaluate the overall metric with the numpy arrays
-    # This ensures consistency in how metric_function is called
-    if s_w is not None:
-        result.overall = metric_function(y_a, y_p, sample_weight=s_w, **kwargs)
-    else:
-        result.overall = metric_function(y_a, y_p, **kwargs)
-
+    result = dict()
     groups = np.unique(group_membership)
     for group in groups:
         group_indices = (group == g_d)
         group_actual = y_a[group_indices]
         group_predict = y_p[group_indices]
         group_weight = None
+        group_key = _BY_GROUP_FORMAT.format(group)
         if s_w is not None:
             group_weight = s_w[group_indices]
-            result.by_group[group] = metric_function(group_actual,
-                                                     group_predict,
-                                                     sample_weight=group_weight,
-                                                     **kwargs)
+            result[group_key] = metric_function(group_actual,
+                                                group_predict,
+                                                sample_weight=group_weight,
+                                                **kwargs)
         else:
-            result.by_group[group] = metric_function(group_actual,
-                                                     group_predict,
-                                                     **kwargs)
+            result[group_key] = metric_function(group_actual,
+                                                group_predict,
+                                                **kwargs)
+
+    # Compute all the statistics, taking care not to stomp on our dictionary
+    try:
+        minimum = min(result.values())
+        argmin = [k for k, v in result.items() if v == minimum]
+        maximum = max(result.values())
+        argmax = [k for k, v in result.items() if v == maximum]
+        result_range = maximum - minimum
+
+        if minimum < 0:
+            range_ratio = np.nan
+        elif maximum == 0:
+            # min==max==0
+            range_ratio = 1
+        else:
+            range_ratio = minimum/maximum
+
+        result[_MIN] = minimum
+        result[_ARGMIN] = argmin
+        result[_MAX] = maximum
+        result[_ARGMAX] = argmax
+        result[_RANGE] = result_range
+        result[_RANGE_RATIO] = range_ratio
```

Comment:
how often are these two used? If not often, the user could compute them from the other values.

Reply:
I would actually be happy to see all of these extra fields eliminated. Since there's no constraint on changing the other members, they could easily be inconsistent. That's why I changed `GroupMetricResult` to computing them dynamically.